Use std::time API instead of u64s for timestamp and expiry time

sgeisler opened the pull request:
Using dedicated time structs should make the API less error prone and easier to use right. Since we don't want to pull in unnecessary Dependencies I didn't use `chrono`.

But `std::time` unfortunately misses a way to do time calculations (like `SystemTime + Duration`) in a non-panicking way (see rust-lang/rust#55527). So I had to make some wild guesses about our potential target platforms and their internal representations of time which can't be statically asserted yet unfortunately. That means we end up with some really ugly heuristics that try to keep people from using the library in ways that would expose them to DoS attacks by overflowing the time calculations.

```rust
// TODO: fix before 2037 (see rust PR #55527)
/// Defines the maximum UNIX timestamp that can be represented as `SystemTime`. This is checked by
/// one of the unit tests, please run them.
const SYSTEM_TIME_MAX_UNIX_TIMESTAMP: u64 = std::i32::MAX as u64;

/// Allow the expiry time to be up to one year. Since this reduces the range of possible timestamps
/// it should be rather low as long as we still have to support 32bit time representations
const MAX_EXPIRY_TIME: u64 = 60 * 60 * 24 * 356;

/// This function is used as a static assert for the size of `SystemTime`. If the crate fails to
/// compile due to it this indicates that your system uses unexpected bounds for `SystemTime`. You
/// can remove this functions and run the test `test_system_time_bounds_assumptions`. In any case,
/// please open an issue. If all tests pass you should be able to use this library safely by just
/// removing this function till we patch it accordingly.
fn __system_time_size_check() {
	/// Use 2 * sizeof(u64) as expected size since the expected underlying implementation is storing
	/// a `Duration` since `SystemTime::UNIX_EPOCH`.
	unsafe { std::mem::transmute::<SystemTime, [u8; 16]>(UNIX_EPOCH); }
}
```

I nonetheless think that it's the right thing to improve the API in that way. With some luck my rust-lang PR will end up in the debian stable rust release in the next decade :sob: but it's ok as long as that will happen before 2037 :smile:

Since @TheBlueMatt is the only user of this library I know of: what do you think about that change. I think that was the last short-term API-breaking change, so I'd really like to publish a `0.1.0` soon.

sgeisler committed:
Use std::time::SystemTime instead of u64 for timestamps

sgeisler committed:
Use std::time::Duration instead of u64 for expiry time

TheBlueMatt commented on README.md:
```diff
@@ -4,6 +4,9 @@
 This repo provides datastructures for BOLT 11 lightning invoices.
 It provides functions to parse and serialize invoices from and to bech32.
 
+**Please be sure to run the test suite since we need to check assumptions
```
Ugh, is it even possible to get clients of clients of the lib to run the test suite? At a minimum it should be exposed so that clients can just call rust-lightning-invoice::check_platform() during their tests.

sgeisler replied:
That sounds like a good idea.

TheBlueMatt commented on src/lib.rs:
```diff
@@ -355,8 +388,8 @@ impl<D: tb::Bool, H: tb::Bool, T: tb::Bool> InvoiceBuilder<D, H, T> {
 	/// Sets the expiry time in seconds.
-	pub fn expiry_time_seconds(mut self, expiry_seconds: u64) -> Self {
-		self.tagged_fields.push(TaggedField::ExpiryTime(ExpiryTime {seconds: expiry_seconds}));
+	pub fn expiry_time(mut self, expiry_time: Duration) -> Self {
```
This should have a doc that indicates it will panic if expiry_time is greater than a year (well, currently MAX_UNIX_TIMESTAMP - a year, but I think thats a bug?), or probably add the doc and then instead of unwrap() store the Err in self.error?

sgeisler replied:
That's a bug, I added proper error handling.

TheBlueMatt commented on src/lib.rs:
```diff
@@ -915,6 +1016,32 @@ impl Deref for PayeePubKey {
 
+impl ExpiryTime {
+	pub fn from_seconds(seconds: u64) -> Result<ExpiryTime, CreationError> {
+		if seconds <= SYSTEM_TIME_MAX_UNIX_TIMESTAMP - MAX_EXPIRY_TIME {
```
Did you mean to just make sure its <= MAX_EXPIRY_TIME?

sgeisler replied:
Yes, at first I used `i64::MAX` as maximum for `SystemTime` and just defined that both `PositiveTimestamp` and `ExpiryTime` could be up to `i64::MAX / 2` big. But then I looked at other platforms and had to decrease the maximum to `i32::MAX` which made dividing by 2 no longer feasible and I had to define two thresholds. 

I probably didn't think much about it and just replaced every `SYSTEM_TIME_MAX_UNIX_TIMESTAMP / 2` with `SYSTEM_TIME_MAX_UNIX_TIMESTAMP - MAX_EXPIRY_TIME` not taking the asymmetry into account.

TheBlueMatt commented on src/lib.rs:
```diff
@@ -915,6 +1016,32 @@ impl Deref for PayeePubKey {
+	}
+
+	pub fn from_duration(duration: Duration) -> Result<ExpiryTime, CreationError> {
+		if duration.as_secs() <= SYSTEM_TIME_MAX_UNIX_TIMESTAMP - MAX_EXPIRY_TIME {
```
Did you mean to just make sure its <= MAX_EXPIRY_TIME?

TheBlueMatt commented on src/lib.rs:
```diff
@@ -717,6 +763,61 @@ impl RawInvoice {
 }
 
+impl PositiveTimestamp {
+
```
nit: stray newline?

TheBlueMatt commented on src/lib.rs:
```diff
@@ -997,9 +1130,23 @@ mod test {
+		/// If this test fails on your platform, please don't use the library and open an issue
+		/// instead so we can resolve the situation. Currently this library is tested on:
+		///   * Linux (64bit)
+		let _ = UNIX_EPOCH + Duration::from_secs(::std::i64::MAX as u64);
```
Is this guaranteed to panic on all platforms? Maybe (UNIX_EPOCH - res).as_secs() on the result and double-check that you get back i64::MAX?

sgeisler replied:
I think it is guaranteed to panic in debug mode if an overflow occurs.

I made the test use the `SYSTEM_TIME_MAX_UNIX_TIMESTAMP` constant instead of `i64::MAX` (since that's our actual assumption) and added some more tests.

TheBlueMatt reviewed (commented):
At a high level I think its nice to use the std::time API just for convenience though I don't think it really helps safety given that we have to restrict the timestamp + expiry delta to make sure they dont overflow anyway (and much better to do it during deserialize than hope the user remembers to avoid the unwrap() either way).

sgeisler commented:
> I don't think it really helps safety given that we have to restrict the timestamp + expiry delta to make sure they dont overflow anyway

I'd argue that it removes possible logic errors where you accidentally pass a completely unrelated integer into a function expecting an expiry time etc. It makes you think twice if the integer you have is actually a duration (and if so with which unit) and should be converted accordingly.

sgeisler commented on src/lib.rs:
```diff
@@ -442,17 +480,30 @@ impl<D: tb::Bool, T: tb::Bool> InvoiceBuilder<D, tb::False, T> {
 	/// Sets the timestamp. `time` is a UNIX timestamp.
-	pub fn timestamp(mut self, time: u64) -> InvoiceBuilder<D, H, tb::True> {
-		self.timestamp = Some(time);
+	pub fn timestamp_raw(mut self, time: u64) -> InvoiceBuilder<D, H, tb::True> {
```
I think the safety argument I made would require me to remove this function since it encourages non-semantic type usage. Some tests will become a bit more painful, but that should be ok.

sgeisler committed:
remove timestamp_raw function from builder

sgeisler committed:
add check_platform function

TheBlueMatt commented on src/lib.rs:
```diff
@@ -43,6 +43,35 @@ fn __system_time_size_check() {
 }
 
 
+/// Call this function on startup to ensure that all assumptions about the platform are valid.
```
nit: Can you document a bit better what they are, and that they are not believed to be violated on any supported rust platforms, and, thus, users should really, really just call this cause its cheap and fast and ensures they're sane.

sgeisler replied:
I extended the function's documentation accordingly. As part of #10 I also added more documentation and would probably add a crate level warning after rebasing.

sgeisler commented:
I guess I'll just merge now, I think I addressed all of your comments. And btw thx for making me implement fuzz tests (not PR ready yet), they already found a round trip/serialization bug in this PR :smiley: (fixed by now)
